### PR TITLE
Multiclass Classification

### PR DESCRIPTION
**Pull request by junseo-park**
I never made a PR, so @SamLau95 didn't have a way to leave comments on the Multiclass Classification outline. Outline available on the "Add multiclass classification" commit.

### CODE REVIEW

File: notebooks/ch17/classification_multiclass.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+
+## Multiclass Classification
+
+Should this be a separate section or just an addendum to an existing section?
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: I think this makes sense as a (short) new section since the previous sections are full of information already.

File: notebooks/ch17/classification_multiclass.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+- What is multiclass classification
+    - More than 2 labels that can be classified; versus a binary classification which is restricted to 2 labels
+- A situation where multiclass classification might be used
+    - Preferably come up with an example in lebron.csv, but might have to use a different data set
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: My guess is that we'll have to use a different dataset. We should ask Deb for suggestions.

File: notebooks/ch17/classification_multiclass.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+- OvO vs OvR
+    - In logistic regression, multiclass classification comes down to a bunch of binary problems
+    - OvO: Fit a binary classification problem to each pair of classes, then vote for class that wins the most "head-to-head" classifications
+    - OvR/OvA: Fit a binary classification problem for each class versus the rest of the classes; result is class that has the highest probability
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: I think we can just talk about OvR since I believe it's more commonly used.

File: notebooks/ch17/classification_multiclass.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+- Example walking through multiclass? sklearn's logistic regression supports OvR out of the box.
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: I think we should do this visually first without code (show multiple decision boundaries and the final point classification), then use sklearn. The more important part is the idea anyway since the code will likely change quickly as sklearn gets updates.

File: notebooks/ch17/classification_multiclass.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+sns.lmplot('$x_1$', '$x_2$', data=shapes, hue='$y$', markers=['^', 's', 'o'], fit_reg=False)
+plt.xlim(1.0, 4.0)
+plt.ylim(1.0, 4.0);
+```
+
+
+![png](classification_multiclass_files/classification_multiclass_5_0.png)
+
+
+Our goal is to build a multiclass classifier that labels observations as `triangle`, `square`, or `circle` given values for $x_1$ and $x_2$. First, we want to build a binary classifier that predicts observations as `triangle`/not `triangle`:
```
Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: > `triangle`/not `triangle`:

nit: I think `triangle` or not `triangle`reads better (just a personal preference)

Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: > we want to build a binary classifier

I think you should mention the name of your classifier `lr_triangle` here (and likewise for `lr_square` and `lr_circle` with their graphs) since we can see the graph and the `plot_binary` function but it isn't clear that an actual LogisticRegression model object is being created.

File: notebooks/ch17/classification_multiclass.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+```python
+plot_binary(shapes, 'circle')
+```
+
+
+![png](classification_multiclass_files/classification_multiclass_10_0.png)
+
+
+We know that the output of the sigmoid function in logistic regression is a probability value from 0 to 1. Then to solve our multiclass classification task, we find the probability of the positive class in each binary classifier and select the class that outputs the highest positive class probability. For example, if we have a new observation with the following values:
```
Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: > Then to solve our multiclass classification task

nit: I think the "Then" can be removed so the sentence starts with "To solve our multiclass..."

File: notebooks/ch17/classification_multiclass.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+3.2   | 2.5
+
+Then our multiclass classifier would fit three binary classifiers to each of the labels of `triangle`, `square`, and `circle`. We extract the positive class probability of each of the three classifiers:
+
+
+```python
+# HIDDEN
+lr_triangle = LogisticRegression(random_state=42)
+lr_triangle.fit(shapes[['$x_1$', '$x_2$']], shapes['$y$'] == 'triangle')
+proba_triangle = lr_triangle.predict_proba([[3.2, 2.5]])[0][1]
```
Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: I think lines 122, 126, and 130 should not be hidden (the ones with `lr_shape.predict_proba`) since this is where the magic happens and it sets up the Iris example well for later.

File: notebooks/ch17/classification_multiclass.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    - OvO: Fit a binary classification problem to each pair of classes, then vote for class that wins the most "head-to-head" classifications
+    - OvR/OvA: Fit a binary classification problem for each class versus the rest of the classes; result is class that has the highest probability
+- Example walking through multiclass? sklearn's logistic regression supports OvR out of the box.
+- Multilabel: More than one classification problem (e.g. a document can be positive/negative, religion/not religion, political/apolitical, etc.)
+
+## Multiclass Classification
+
+So far we have discussed binary classification, in which our classifier determines whether an observation is part of the positive class (e.g. an email is classified as spam) or the negative class (e.g. an email is not spam). However, many data science problems involve **multiclass classification**, in which we would like to classify observations as one of several different classes. For example, we may be interested in classifying emails into folders such as Family, Friends, Work, and Promotions. To solve these types of problems, we use a new method called **one-vs-rest (OvR) classification**.
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: I think the first sentence is simpler as:

> Our classifiers thus far perform binary classification where each observation is one of two classes; we classified emails as either ham or spam, for example.

File: notebooks/ch17/classification_multiclass.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+    [[1.3, 3.6, 'triangle'], [1.6, 3.2, 'triangle'], [1.8, 3.8, 'triangle'],
+     [2.0, 1.2, 'square'], [2.2, 1.9, 'square'], [2.6, 1.4, 'square'],
+     [3.2, 2.9, 'circle'], [3.5, 2.2, 'circle'], [3.9, 2.5, 'circle']],
+    columns=['$x_1$', '$x_2$', '$y$']
+)
+```
+
+
+```python
+sns.lmplot('$x_1$', '$x_2$', data=shapes, hue='$y$', markers=['^', 's', 'o'], fit_reg=False)
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: nit: Add a `# HIDDEN` here.

File: notebooks/ch17/classification_multiclass.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+```python
+# HIDDEN
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import seaborn as sns
+from sklearn.datasets import load_iris
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import confusion_matrix
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: You can add:

```
pd.options.display.max_rows = 7
pd.options.display.max_columns = 8
pd.set_option('precision', 2)
```

Here to make DFs only display 2 numbers after the decimal and have better DF displaying.

(See https://github.com/DS-100/textbook/blob/master/starter.ipynb for the starter code I recommend.)

File: notebooks/ch17/classification_multiclass.md [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+lr_square.fit(shapes[['$x_1$', '$x_2$']], shapes['$y$'] == 'square')
+proba_square = lr_square.predict_proba([[3.2, 2.5]])[0][1]
+
+lr_circle = LogisticRegression(random_state=42)
+lr_circle.fit(shapes[['$x_1$', '$x_2$']], shapes['$y$'] == 'circle')
+proba_circle = lr_circle.predict_proba([[3.2, 2.5]])[0][1]
+
+print('Triangle:', proba_triangle)
+print('Square:', proba_square)
+print('Circle:', proba_circle)
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: Might as well omit the print statements if we're displaying the table right after.

File: notebooks/ch17/classification_multiclass.md [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+
+
+    LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
+              intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
+              penalty='l2', random_state=42, solver='liblinear', tol=0.0001,
+              verbose=0, warm_start=False)
+
+
+
+We predict on the test data, and use a confusion matrix to evaluate the results:
```
Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: I would plot the confusion matrix with the labels and both axes specified (you can find the code in lab 11 or my sensitivity/specificity section).

File: notebooks/ch17/classification_multiclass.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+![](petal_sepal.png)
+
+We will create a multiclass classifier that predicts the type of Iris plant based on the four features above. First, we read in the data:
+
+
+```python
+iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data',
+                  header=None, names=['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species'])
+
+iris.head()
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: With the additional pandas config above, you can just display `iris` for a nicer output (it shows the total number of rows too).

File: notebooks/ch17/classification_multiclass.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+
+
+
+    array([[19,  0,  0],
+           [ 0, 15,  2],
+           [ 0,  0, 17]])
+
+
+
+The confusion matrix shows that our classifier misclassified two `Iris-versicolour` observations as `Iris-virginica`. In the real world, certain misclassifications can be more common than others. Confusion matrices are valuable because they help us identify where our classifier struggles, and thus provides insight on what kind of features we may need to extract in order to improve the classifier.
```
Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: It would be helpful to briefly mention why certain misclassifications can be more common than others (what properties to these share vs. other less frequent misclassifications?)

File: notebooks/ch17/classification_multiclass.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+
+
+
+We predict on the test data, and use a confusion matrix to evaluate the results:
+
+
+```python
+# to-do: add x-axis, y-axis labels to confusion matrix
+y_pred = lr.predict(X_test)
+confusion_matrix(y_test, y_pred)
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: Can you use https://seaborn.pydata.org/generated/seaborn.heatmap.html to display a heatmap of the matrix with the axis labels and such? It'll be easier to interpret.

File: notebooks/ch17/classification_multiclass.md (hunk 14 above)
Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: > where our classifier struggles

Instead of struggles, perhaps something along the lines of "help us identify the errors our classifier makes"

File: notebooks/ch17/classification_multiclass.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+           [ 0, 15,  2],
+           [ 0,  0, 17]])
+
+
+
+The confusion matrix shows that our classifier misclassified two `Iris-versicolour` observations as `Iris-virginica`. In the real world, certain misclassifications can be more common than others. Confusion matrices are valuable because they help us identify where our classifier struggles, and thus provides insight on what kind of features we may need to extract in order to improve the classifier.
+
+## Multilabel Classification
+
+Another type of classification problem is **multilabel classification**, in which each observation can have multiple labels. An example would be a document classification system: a document can be classified as having positive or negative sentiments, but it can also be distinguished between religious/nonreligious or political/apolitical. Multilabel problems can also be multiclass; we may want our document classification system to distinguish between a list of genres, as well as identify the language that the document is written in.
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: The second sentence might be better as:

> Consider a system for classifying documents: a single document can have positive or negative sentiment, religious or nonreligious content, and liberal or conservative political leaning.

File: notebooks/ch17/classification_multiclass.md [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+
+## Multilabel Classification
+
+Another type of classification problem is **multilabel classification**, in which each observation can have multiple labels. An example would be a document classification system: a document can be classified as having positive or negative sentiments, but it can also be distinguished between religious/nonreligious or political/apolitical. Multilabel problems can also be multiclass; we may want our document classification system to distinguish between a list of genres, as well as identify the language that the document is written in. 
+
+Because each set of labels in a multiclass classification is mutually exclusive, each problem can be solved independently of the others. We can then combine the outputs so that every observation is assigned a set of classification labels.
```
Comment (SamLau95) [inline notebooks/ch17/classification_multiclass.md]: We can be a bit more explicit here:

> We may perform multilabel classification by simply training a separate classifier on each set of labels. To label a new point, we combine each classifier's predictions.

Comment (ananthagarwal) [inline notebooks/ch17/classification_multiclass.md]: (Just adding to Sam's comment here): Consider the definition given here also http://scikit-learn.org/stable/modules/multiclass.html

File: notebooks/ch17/classification_multiclass.md (hunk 11 above)
Comment (ananthagarwal) [reply notebooks/ch17/classification_multiclass.md]: Personally I feel it would be beneficial to show only the three `predict_proba` lines of the cell, keep the print statements, and omit the DataFrame, but this works too!

Comment (junseo-park) [reply notebooks/ch17/classification_multiclass.md]: I think in this snippet, it's better to introduce just the theory behind OvR (and thus show just the table and not the `predict_proba`). In practice, students probably won't be running `predict_proba` on each class and then selecting the class with the highest output; it might be clearer to the reader to let the only code implementations shown in this section to be the default OvR implementation in LogisticRegression (shown later in the Iris example).

File: notebooks/ch17/classification_multiclass.md (hunk 15 above)
Comment (junseo-park) [reply notebooks/ch17/classification_multiclass.md]: The output didn't show up in the markdown file, but the confusion matrix looks like this:
image
So the seaborn heatmaps don't show up well:
image
I think keeping it as a table with numbers and axis labels might be easier to interpret. What do you think?

(I found this pretty funny, actually; I had to deliberately set the test size to 0.35 in order to get any misclassifications at all. The dataset doesn't pose much challenge to logistic regression, unfortunately.)

Comment (ananthagarwal) [reply notebooks/ch17/classification_multiclass.md]: Yeah I ran into a similar problem with the spam/ham dataset, so I increased the number of data points (which was possible since the email database is huge) and tried different model features until I got one that would be useful for instructional purposes. I don't think either of these are possible for the iris data (considering all features are given), so this is probably fine. Not sure how the `sns.heatmap` colors work but I think blue is a good color.

File: notebooks/ch17/classification_multiclass.md [hunk 18]
```diff
@@ -281,32 +290,40 @@ lr.fit(X_train, y_train)
...
+Another type of classification problem is **multilabel classification**, in which each observation can have multiple labels. An example would be a document classification system: a document can have positive or negative sentiment, religious or nonreligious content, and liberal or conservative leaning. Multilabel problems can also be multiclass; we may want our document classification system to distinguish between a list of genres, or identify the language that the document is written in. 
 
-Because each set of labels in a multiclass classification is mutually exclusive, each problem can be solved independently of the others. We can then combine the outputs so that every observation is assigned a set of classification labels.
+Because each set of labels in a multilabel classification is mutually exclusive, we may perform multilabel classification by simply training a separate classifier on each set of labels. To label a new point, we combine each classifier's predictions.
```
Comment (junseo-park) [inline notebooks/ch17/classification_multiclass.md]: This is a mouthful; I'll take out the dependent clause and start the sentence with "We may perform..." in the next commit.

File: notebooks/ch17/classification_multiclass.md [hunk 19]
```diff
@@ -281,32 +290,40 @@ lr.fit(X_train, y_train)
...
 
-We predict on the test data, and use a confusion matrix to evaluate the results:
+We predict on the test data, and use a confusion matrix to evaluate the results.
 
 
 ```python
-# to-do: add x-axis, y-axis labels to confusion matrix
 y_pred = lr.predict(X_test)
-confusion_matrix(y_test, y_pred)
+plot_confusion_matrix(y_test, y_pred)
```
Comment (junseo-park) [inline notebooks/ch17/classification_multiclass.md]: I'm not a huge fan of how the diagonal has different shades even though all observations on the diagonal are correctly classified. Do you think this is okay as is, should we normalize (and then lose the annotations about number of observations), or should we manually create a heatmap in `matplotlib` with the diagonals set to black?
